core-bundle/src/Routing/ResponseContext/HtmlHeadBag/HtmlHeadBag.php

```diff
@@ -89,11 +91,20 @@ public function getCanonicalUri(): string
         return $this->canonicalUri;
     }
 
-    public function getCanonicalUriForRequest(Request $request): string
+    public function getCanonicalUriForRequest(Request $request, InsertTagParser|null $insertTagParser = null): string
```

Comment:
Rather than having to pass the `InsertTagParser` instance yourself, wouldn't it be better to inject that into the `HtmlHeadBag` instance via its own `HtmlHeadBagFactory` service? Like @Toflar originally suggested.

Reply:
I don't think so. We're also not passing the request stack. It would also change the constructor of the service, which in my optionion would be a bad idea since other request handlers could initialize that object as well.

Reply:
I'm really not sure here. I think there are good arguments for both approaches so I'm marking this up for discussion.

Reply:
I have reverted the implementation and simply duplicated the current page code in 7e11216514d236ca9ad6e16512c47182185e5656. We can refactor that once we have https://github.com/contao/contao/pull/6547#issuecomment-1825263152